### PR TITLE
Client SDK changes for foundation based bare metal provisioning

### LINKED ISSUE #354: [Foundation] Provision Image Nodes
Nutanix provides deployment of nutanix cluster via Foundation. 

Add capability to deploy image nodes or clusters via terraform provider.

APIs: https://www.nutanix.dev/api_references/foundation/#/ZG9jOjQ1Mg-foundation-hub

### PR DESCRIPTION
**Pull request by bhatipradeep**
Changes:

- Added a new foundation client with appropriate structs.
- Updated(Temporary) client_test.go as per changes in credentials struct, which now also have foundation related attributes.
- Updated config.go and provider.go for accomodating newly created foundation client and its requirements.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -20,6 +20,7 @@ import (
 const (
 	// libraryVersion = "v3"
 	defaultBaseURL = "https://%s/"
+	httpBaseURL    = "http://%s/"
```
Comment (siddharth-nutanix) [inline client/client.go]: Can we use same base URL and make http / https dynamic by using boolean flag isHTTP?

File: client/foundation/foundation_file_management_service.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	ListNOSPackages() (*ListNOSPackagesResponse, error)
+	ListHypervisorISOs() (*ListHypervisorISOsResponse, error)
+}
+
+type FileManagementOperations struct {
+	client *client.Client
+}
+
+//Lists the available AOS packages in Foundation
+func (fileManagementOperations FileManagementOperations) ListNOSPackages() (*ListNOSPackagesResponse, error) {
```
Comment (siddharth-nutanix) [inline client/foundation/foundation_file_management_service.go]:
```suggestion
func (fmo FileManagementOperations) ListNOSPackages() (*ListNOSPackagesResponse, error) {
```
Use smaller variable names as we already have type defined.

Comment (siddharth-nutanix) [reply client/foundation/foundation_file_management_service.go]: Make changes everywhere.

File: client/foundation/foundation_file_management_service.go [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+	ctx := context.TODO()
```
Comment (siddharth-nutanix) [inline client/foundation/foundation_file_management_service.go]: pass context everywhere.

Comment (siddharth-nutanix) [reply client/foundation/foundation_file_management_service.go]: make changes everywhere.

File: nutanix/config.go [hunk 4]
```diff
@@ -13,27 +14,32 @@ const Version = "3.1"
...
+		Endpoint:           c.Endpoint,
+		Username:           c.Username,
+		Password:           c.Password,
+		Port:               c.Port,
+		Insecure:           c.Insecure,
+		SessionAuth:        c.SessionAuth,
+		ProxyURL:           c.ProxyURL,
+		FoundationEndpoint: c.FoundationEndpoint,
+		FoundationPort:     c.FoundationPort,
+		FoundationURL:      fmt.Sprintf("%s:%s", c.FoundationEndpoint, c.FoundationPort),
```
Comment (siddharth-nutanix) [inline nutanix/config.go]: Do we need FoundationURL as separate member of client.Credentials ?
Whenever we need the foundation URL we can always create it using endpoint and port.

File: nutanix/provider.go [hunk 5]
```diff
@@ -28,6 +28,10 @@ func Provider() *schema.Provider {
 			"note, this is never the data services VIP, and should not be an\n" +
 			"individual CVM address, as this would cause calls to fail during\n" +
 			"cluster lifecycle management operations, such as AOS upgrades.",
+
+		"foundation_endpoint": "URL for foundation VM (eg. Foundation VM IP)",
```
Comment (siddharth-nutanix) [inline nutanix/provider.go]:
```suggestion
		"foundation_endpoint": "endpoint for foundation VM (eg. Foundation VM IP)",
```

**Review by siddharth-nutanix — CHANGES_REQUESTED** [review]

File: client/foundation/foundation_api.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+)
+
+//Foundation client with its services
+type Client struct {
+
+	//base client
+	client *client.Client
+
+	//Service for Imaging Nodes and Cluster Creation
+	NodeImaging NodeImagingService
```
[the remark this answers is not in the document]
Comment (bhatipradeep) [reply client/foundation/foundation_api.go:21]: Done

File: client/client.go [hunk 7]
```diff
@@ -96,7 +98,12 @@ func NewClient(credentials *Credentials, userAgent string, absolutePath string)
 
 	httpClient.Transport = logging.NewTransport("Nutanix", transCfg)
 
-	baseURL, err := url.Parse(fmt.Sprintf(defaultBaseURL, credentials.URL))
+	protocol := "https"
```
Comment (siddharth-nutanix) [inline client/client.go:101]: lets move http / https in constants.

Comment (siddharth-nutanix) [reply client/client.go:101]: You can also create a method to figure out and construct a baseURL

**Review by siddharth-nutanix — APPROVED** [review]

**Pull request merged**